File: rectangular_channel/elm_arc.f
      subroutine elm_arc
     +
     +  (N
     +  ,ratio
     +  ,Xcnt,Ycnt
     +  ,radius
     +  ,angle1,angle2
     +  ,sinit
     +  ,Isym
     +  ,Xe,Ye,Te,se
     +  ,Xm,Ym,Tm,sm
     +  )

c=============================================
c FDLIB, BEMLIB
c
c Copyright by C. Pozrikidis, 1999
c All rights reserved
c
c This program is to be used only under the
c stipulations of the licensing agreement
c=============================================

c----------------------------------------------------------------
c  Disretization of a circular arc into N elements
c
c  SYMBOLS:
c  -------
c
c  Xcnt, Ycnt:	center of the arc
c  ratio:	radius of the arc
c
c  ratio: If Isym = 0, ratio of length of LAST to FIRST element
c         If Isym = 1, ratio of length of MID  to FIRST element
c
c  Xe, Ye, Te:	coordinates and polar angle for the end-nodes
c  Xm, Ym, Tm:	coordinates and polar angle for the mid-nodes
c
c  sinit: assigned arc length at the first point
c
c----------------------------------------------------------------

      Implicit Double Precision (a-h,o-z)

      Dimension Xe(129),Ye(129),Te(129),se(129)
      Dimension Xm(128),Ym(128),Tm(128),sm(128)

c------------
c one element
c------------

      If(N.eq.1) then

       te(1) = angle1;
       xe(1) = xcnt + radius*cos(te(1))
       ye(1) = ycnt + radius*sin(te(1))
       se(1) = sinit

       te(2) = angle2
       xe(2) = xcnt + radius*Dcos(te(2))
       ye(2) = ycnt + radius*Dsin(te(2))
       se(2) = se(1)+radius*abs(te(2)-te(1))

       Go to 99

      End If

c--------
c prepare
c--------

      N1 = N+1

c---------------------------
c non-symmetric distribution
c---------------------------

      If(Isym.eq.0) then

      texp  = 1.0D0/(N-1.0D0)
      alpha = ratio**texp
 
      If(Dabs(alpha-1.0D0).gt.0.0000001) then
       factor = (1.0D0-alpha)/(1.0D0-alpha**N)
      Else
       factor = 1.0D0/(N-1.0D0+1.0D0)
      End If

      deltat = (angle2-angle1)*factor   ! aperture of first element

      te(1) = angle1                      ! first point
      Xe(1) = Xcnt + radius*Dcos(te(1))
      Ye(1) = Ycnt + radius*Dsin(te(1))
      se(1) = sinit

      Do i=2,N1
        Te(i) = Te(i-1)+deltat
        Xe(i) = Xcnt + radius*Dcos(Te(i))
        Ye(i) = Ycnt + radius*Dsin(Te(i))
        se(i) = se(i-1)+radius*Dabs(deltat)
        deltat = deltat*alpha
      End Do

      Go to 99

      End If

c-----------------------
c symmetric distribution
c----------------------

c---
      If(N.eq.2) then
c---

      te(1) = angle1;
      xe(1) = xcnt + radius*cos(te(1))
      ye(1) = ycnt + radius*sin(te(1))
      se(1) = sinit

      angleh = 0.5D0*(angle1+angle2)     ! mid-point
      te(2) = angleh
      xe(2) = xcnt + radius*cos(te(2))
      ye(2) = ycnt + radius*sin(te(2))
      se(2) = se(1)+radius*abs(te(2)-te(1))

      te(3) = angle2
      xe(3) = xcnt + radius*cos(te(3))
      ye(3) = ycnt + radius*sin(te(3))
      se(3) = se(2)+radius*abs(te(3)-te(2))

      Go to 99

c----
      End If
c----

c----
      If(mod(N,2).eq.0) then  ! even number of elements
c----

      angleh = 0.5D0*(angle1+angle2)      ! mid-point

      Nh  = N/2
      Nh1 = Nh+1

      texp  = 1.0D0/(Nh-1.0D0)
      alpha = ratio**texp

      If(abs(alpha-1.0D0).gt.0.0000001) then
       factor = (1.0D0-alpha)/(1.0D0-alpha**Nh)
      Else
       factor = 1.0D0/(Nh-1.0D0+1.0D0)
      End If

      deltat = (angleh-angle1) * factor   ! aperture of first element

      te(1) = angle1                      ! first point
      Xe(1) = Xcnt + radius*Dcos(te(1))
      Ye(1) = Ycnt + radius*Dsin(te(1))
      se(1) = sinit

      Do i=2,Nh1                          ! up to mid-point
        te(i) = te(i-1)+deltat
        Xe(i) = Xcnt + radius*Dcos(te(i))
        Ye(i) = Ycnt + radius*Dsin(te(i))
        se(i) = se(i-1)+radius*Dabs(deltat)
        deltat = deltat*alpha
      End Do

      deltat = deltat/alpha

      Do i=Nh1+1,N1                        ! reflect
        te(i) = te(i-1)+deltat
        Xe(i) = Xcnt + radius*Dcos(te(i))
        Ye(i) = Ycnt + radius*Dsin(te(i))
        se(i) = se(i-1)+radius*Dabs(deltat)
        deltat = deltat/alpha
      End Do

      Go to 99

      End If

c-----------------------
c odd number of elements
c-----------------------

      texp  = 2.0D0/(N-1.0D0)
      alpha = ratio**texp

      If(Dabs(alpha-1.0D0).gt.0.0000001) then
       tmp1   = (N+1.0)/2.0D0
       tmp2   = (N-1.0)/2.0D0
       factor = (1.0D0-alpha)/(2.0D0-alpha**tmp1-alpha**tmp2)
      Else
       factor = 1.0/N
      End If

      deltat = (angle2-angle1) * factor   ! aperture of first element

      te(1) = angle1                      ! first point
      Xe(1) = Xcnt + radius*Dcos(te(1))
      Ye(1) = Ycnt + radius*Dsin(te(1))
      se(1) = sinit

      Do i=2,(N+3)/2                      ! up to mid point + 1
        Te(i) = Te(i-1)+deltat
        Xe(i) = Xcnt + radius*Dcos(te(i))
        Ye(i) = Ycnt + radius*Dsin(te(i))
        se(i) = se(i-1)+radius*Dabs(deltat)
        deltat = deltat*alpha
      End Do

      deltat = deltat/(alpha**2)

      Do i=(N+5)/2,N1
        Te(i) = Te(i-1)+deltat
        Xe(i) = Xcnt + radius*Dcos(te(i))
        Ye(i) = Ycnt + radius*Dsin(te(i))
        se(i) = se(i-1)+radius*Dabs(deltat)
        deltat = deltat/alpha
      End Do

c-----------------------------------

  99  Continue

c-----------------------
c compute the mid-points
c-----------------------

      Do i=1,N
       tm(i) = 0.5D0*(te(i)+te(i+1))
       sm(i) = 0.5D0*(se(i)+se(i+1))
       Xm(i) = Xcnt + radius*Dcos(tm(i))
       Ym(i) = Ycnt + radius*Dsin(tm(i))
      End Do

c-----
c Done
c-----

      Return
      End

File: rectangular_channel/elm_arc.m
function [xe,ye,te,se,xm,ym,tm,sm]...
 ...
 = elm_arc (N,ratio,xcnt,ycnt ...
           ,radius,angle1,angle2 ...
           ,sinit,Isym)

%-----------------------------------------
% FDLIB, BEMLIB
%
% Copyright by C. Pozrikidis, 1999
% All rights reserved.
%
% This program is to be used only under the
% stipulations of the licensing agreement
%----------------------------------------

%----------------------------------------------------------------
%  Disretization of a circular arc into N elements
%
%  SYMBOLS:
%  -------
%
%  xcnt, ycnt:	center of the arc
%  ratio:	radius of the arc
%
%  ratio: if Isym = 0, ratio of length of LAST to FIRST element
%         if Isym = 1, ratio of length of MID  to FIRST element
%
%  xe, ye, te:	coordinates and polar angle for the end-nodes
%  xm, ym, tm:	coordinates and polar angle for the mid-nodes
%
%  sinit: assigned arc length at the first point
%----------------------------------------------------------------

%------------
% one element
%------------

if(N==1)

 te(1) = angle1; 
 xe(1) = xcnt + radius*cos(te(1));
 ye(1) = ycnt + radius*sin(te(1));
 se(1) = sinit;

 te(2) = angle2;
 xe(2) = xcnt + radius*cos(te(2));
 ye(2) = ycnt + radius*sin(te(2));
 se(2) = se(1)+radius*abs(te(2)-te(1));

%---------------------------
% non-symmetric distribution
%---------------------------

elseif(Isym==0)

   texp  = 1.0/(N-1.0);
   alpha = ratio^texp;
 
    if(abs(alpha-1.0)>0.0000001)
       factor = (1.0-alpha)/(1.0-alpha^N);
    else
       factor = 1.0/N;
    end

    deltat = (angle2-angle1)*factor;  % aperture of first element

      te(1) = angle1;                     % first point
      xe(1) = xcnt + radius*cos(te(1));
      ye(1) = ycnt + radius*sin(te(1));
      se(1) = sinit;

      for i=2:N+1
        te(i) = te(i-1)+deltat;
        xe(i) = xcnt + radius*cos(te(i));
        ye(i) = ycnt + radius*sin(te(i));
        se(i) = se(i-1)+radius*abs(deltat);
        deltat = deltat*alpha;
      end

%-----------------------
% symmetric distribution
%----------------------

  elseif(Isym==1)

%---
   if(N==2)
%---

   te(1) = angle1; 
   xe(1) = xcnt + radius*cos(te(1));
   ye(1) = ycnt + radius*sin(te(1));
   se(1) = sinit;

   angleh = 0.5*(angle1+angle2);     % mid-point
   te(2) = angleh;
   xe(2) = xcnt + radius*cos(te(2));
   ye(2) = ycnt + radius*sin(te(2));
   se(2) = se(1)+radius*abs(te(2)-te(1));

   te(3) = angle2;
   xe(3) = xcnt + radius*cos(te(3));
   ye(3) = ycnt + radius*sin(te(3));
   se(3) = se(2)+radius*abs(te(3)-te(2));

%---
  elseif(mod(N,2)==0) % even number of elements
%---

      angleh = 0.5*(angle1+angle2);     % mid-point

      Nh  = N/2;
      Nh1 = Nh+1;

      texp  = 1.0/(Nh-1.0);
      alpha = ratio^texp;

      if(abs(alpha-1.0)>0.0000001)
       factor = (1.0-alpha)/(1.0-alpha^Nh);
      else
       factor = 1.0/Nh;
      end

      deltat = (angleh-angle1)*factor;  % aperture of first element

      te(1) = angle1;                     % first point
      xe(1) = xcnt + radius*cos(te(1));
      ye(1) = ycnt + radius*sin(te(1));
      se(1) = sinit;

      for i=2:Nh1                          % up to mid-point
        te(i) = te(i-1)+deltat;
        xe(i) = xcnt + radius*cos(te(i));
        ye(i) = ycnt + radius*sin(te(i));
        se(i) = se(i-1)+radius*abs(deltat);
        deltat = deltat*alpha;
      end

      deltat = deltat/alpha;

      for i=Nh1+1:N+1                        % reflect
        te(i) = te(i-1)+deltat;
        xe(i) = xcnt + radius*cos(te(i));
        ye(i) = ycnt + radius*sin(te(i));
        se(i) = se(i-1)+radius*abs(deltat);
        deltat = deltat/alpha;
      end

%---
 else
%---

% odd number of elements

      texp  = 2.0/(N-1.0);
      alpha = ratio^texp;

      if(Dabs(alpha-1.0)>0.0000001)
       tmp1   = (N+1.0)/2.0;
       tmp2   = (N-1.0)/2.0;
       factor = (1.0-alpha)/(2.0-alpha^tmp1-alpha^tmp2);
      else
       factor = 1.0/N;
      end

      deltat = (angle2-angle1)*factor;  % aperture of first element

      te(1) = angle1;                     % first point
      xe(1) = xcnt + radius*cos(te(1));
      ye(1) = ycnt + radius*sin(te(1));
      se(1) = sinit;

      for i=2:(N+3)/2                      % up to mid point + 1
        te(i) = te(i-1)+deltat;
        xe(i) = xcnt + radius*cos(te(i));
        ye(i) = ycnt + radius*sin(te(i));
        se(i) = se(i-1)+radius*abs(deltat);
        deltat = deltat*alpha;
      end

      deltat = deltat/(alpha^2);

      for i=(N+5)/2:N+1
        te(i) = te(i-1)+deltat;
        xe(i) = xcnt + radius*cos(te(i));
        ye(i) = ycnt + radius*sin(te(i));
        se(i) = se(i-1)+radius*abs(deltat);
        deltat = deltat/alpha;
      end

%---
   end
%---

%---
end
%---

%------------------------------
% compute the element midpoints
%------------------------------

for i=1:N
  tm(i) = 0.5*(te(i)+te(i+1));
  sm(i) = 0.5*(se(i)+se(i+1));
  xm(i) = xcnt + radius*cos(tm(i));
  ym(i) = ycnt + radius*sin(tm(i));
end

%-----
% done
%-----

return
